File: tlg/src/main/kotlin/tlg/context/PlatformContext.kt
package tlg.context

import tlg.geometry.TlgPoint
import tlg.geometry.TlgRectangle
import java.io.*

abstract class PlatformContext {
    fun drawRectangle(c: Int, r: TlgRectangle) {
        drawLine(c, r.tL, r.tR)
        drawLine(c, r.tR, r.bR)
        drawLine(c, r.bR, r.bL)
        drawLine(c, r.bL, r.tL)
    }

    abstract fun drawLine(color: Int, p1: TlgPoint, p2: TlgPoint)
    abstract fun drawFilledRectangle(color: Int, rect: TlgRectangle)

    abstract fun colorBackground(): Int
    abstract fun colorDark(): Int
    abstract fun colorHighlight(): Int
    abstract fun colorGrayed(): Int
    abstract fun colorFrame(): Int
    abstract fun colorGrid(): Int
    abstract fun countOfColor(): Int
    abstract fun getColor(i: Int): Int

    //////
    open val configDirectory: File
        get() = File(System.getProperty("user.home"), ".config/tlg")

    @Throws(FileNotFoundException::class)
    fun getInputStream(fileName: String): BufferedInputStream {
        val file = File(configDirectory, fileName)
        return BufferedInputStream(FileInputStream(file))
    }

    @Throws(FileNotFoundException::class)
    fun getOutputStream(fileName: String): BufferedOutputStream {
        val directory = configDirectory
        directory.mkdirs()
        val file = File(directory, fileName)
        return BufferedOutputStream(FileOutputStream(file))
    }

    open fun setDirtyRect(rect: TlgRectangle) {

    }
}

File: js/tlg-lib/src/context/PlatformContext.ts
import { TlgRectangle } from '../tlg.matrix/TlgRectangle'
import { TlgPoint } from '../tlg.matrix/TlgPoint'

export abstract class PlatformContext {
    public drawRectangle (c:number, r:TlgRectangle) {
        this.drawLine(c, r.getTL(), r.getTR())
        this.drawLine(c, r.getTR(), r.getBR())
        this.drawLine(c, r.getBR(), r.getBL())
        this.drawLine(c, r.getBL(), r.getTL())
    }

    public abstract drawLine(color:number, p1:TlgPoint, p2:TlgPoint):void;
    public abstract drawFilledRectangle(color:number, rect:TlgRectangle):void;

    public abstract colorBackground():number;
    public abstract colorDark():number;
    public abstract colorHighlight():number;
    public abstract colorGrayed():number;
    public abstract colorFrame():number;
    public abstract colorGrid():number;

    public abstract countOfColor():number;
    public abstract getColor(i:number):number;
}
